django/db/migrations/optimizer.py

```diff
@@ -44,20 +44,29 @@ def optimize_inner(self, operations, app_label=None):
         """
         new_operations = []
         for i, operation in enumerate(operations):
+            right = True  # Should we reduce on the right or on the left.
             # Compare it to each operation after it
             for j, other in enumerate(operations[i + 1:]):
                 in_between = operations[i + 1:i + j + 1]
                 result = operation.reduce(other, in_between, app_label)
                 if isinstance(result, list):
-                    # Optimize! Add result, then remaining others, then return
-                    new_operations.extend(result)
-                    new_operations.extend(in_between)
+                    if right:
+                        new_operations.extend(in_between)
+                        new_operations.extend(result)
+                    elif all(op.reduce(other, [], app_label) is True for op in in_between):
+                        # Perform a left reduction if all the in between
+                        # operations can optimize through other.
+                        new_operations.extend(result)
+                        new_operations.extend(in_between)
+                    else:
+                        # Else continue trying.
+                        new_operations.append(operation)
+                        break
                     new_operations.extend(operations[i + j + 2:])
                     return new_operations
-                if not result:
-                    # We can't optimize across `other`.
-                    new_operations.append(operation)
-                    break
+                elif not result:
+                    # Can't perform a right reduction.
+                    right = False
```

Comment:
@rtpg this is where the magic happens. We first try to optimize right (that is right of the `in_between`) if that doesn't work we try to optimize left iff all of the operations in between can optimize through other. I'll try to adjust the comments this week end.